python/lsst/daf/butler/datastores/webdavDatastore.py

```diff
@@ -0,0 +1,271 @@
+        location, formatter = self._prepare_for_put(inMemoryDataset, ref)
+
+        if webdavCheckFileExists(location, session=self.session,)[0]:
+            raise FileExistsError(f"Cannot write file for ref {ref} as "
+                                  f"output file {location.uri} exists.")
+
+        try:
+            serializedDataset = formatter.toBytes(inMemoryDataset)
+            self.session.put(location.uri.geturl(), data=serializedDataset)
+            log.debug("Wrote file directly to %s", location.uri)
+        except NotImplementedError:
+            with tempfile.NamedTemporaryFile(suffix=location.getExtension()) as tmpFile:
+                formatter._fileDescriptor.location = Location(*os.path.split(tmpFile.name))
+                formatter.write(inMemoryDataset)
+                with open(tmpFile.name, 'rb') as f:
+                    self.session.put(location.uri.geturl(), data=f)
+                log.debug("Wrote file to %s via a temporary directory.", location.uri)
+
+        if self._transaction is None:
+            raise RuntimeError("Attempting to write artifact without transaction enabled")
+
+        self._transaction.registerUndo("write", webdavDeleteFile, location)
+
+        return self._extractIngestInfo(location.uri.geturl(), ref, formatter=formatter)
+
+    def _overrideTransferMode(self, *datasets: Any, transfer: Optional[str] = None) -> Optional[str]:
+        if transfer != "auto":
+            return transfer
+        return "copy"
+
+    def _standardizeIngestPath(self, path: str, *, transfer: Optional[str] = None) -> str:
+        if transfer not in (None, "move", "copy"):
+            raise NotImplementedError(f"Transfer mode {transfer} not supported.")
+        srcUri = ButlerURI(path)
+        if srcUri.scheme == 'file' or not srcUri.scheme:
+            if not os.path.exists(srcUri.ospath):
+                raise FileNotFoundError(f"File at '{srcUri}' does not exist.")
+        elif srcUri.scheme.startswith("http"):
+            if not webdavCheckFileExists(srcUri, session=self.session)[0]:
+                raise FileNotFoundError(f"File at '{srcUri}' does not exist.")
+        else:
+            raise NotImplementedError(f"Scheme type {srcUri.scheme} not supported.")
+
+        if transfer is None:
+            rootUri = ButlerURI(self.root)
+            if srcUri.scheme == "file":
+                raise RuntimeError(f"'{srcUri}' is not inside repository root '{rootUri}'. "
+                                   "Ingesting local data to WebdavDatastore without upload "
+                                   "to Webdav is not allowed.")
+            elif srcUri.scheme.startswith("http"):
+                if not srcUri.path.startswith(rootUri.path):
+                    raise RuntimeError(f"'{srcUri}' is not inside repository root '{rootUri}'.")
+        return path
+
+    def _extractIngestInfo(self, path: Union[str, ButlerURI], ref: DatasetRef, *,
```

Comment:
This whole method needs to be rewritten following the recent cleanups in s3datastore.py to use ButlerURI everywhere. It's probably the case that it's identical to the s3 one except for the existence check right at the end.